src/framebuffer/core.rs

```diff
@@ -71,127 +109,119 @@ impl<'a> framebuffer::FramebufferBase<'a> for Framebuffer<'a> {
         var_screen_info.vmode = 0; // FB_VMODE_NONINTERLACED
         var_screen_info.accel_flags = 0;
 
-        Framebuffer::put_var_screeninfo(&device, swtfb_client.as_ref(), &mut var_screen_info);
+        let fix_screen_info = match &framebuffer_update {
+            FramebufferUpdate::Ioctl(device) => {
+                Framebuffer::put_var_screeninfo(device, &mut var_screen_info);
+                Framebuffer::get_fix_screeninfo(device)
+            }
+            FramebufferUpdate::Swtfb(c) => c.get_fix_screeninfo(),
+        };
 
-        let fix_screen_info = Framebuffer::get_fix_screeninfo(&device, swtfb_client.as_ref());
         let frame_length = (fix_screen_info.line_length * var_screen_info.yres) as usize;
 
-        let mem_map = if let Some(mem_map) = mem_map {
-            mem_map
-        } else {
-            MmapOptions::new()
+        let mem_map = match &framebuffer_update {
+            FramebufferUpdate::Ioctl(device) => MmapOptions::new()
                 .len(frame_length)
-                .map_raw(&device)
-                .expect("Unable to map provided path")
+                .map_raw(device)
+                .expect("Unable to map provided path"),
+            FramebufferUpdate::Swtfb(swtfb_client) => {
+                let (_, mem_map) = swtfb_client
+                    .open_buffer()
+                    .expect("Failed to open swtfb shared buffer");
+                mem_map
+            }
         };
 
         // Load the font
         let font_data = include_bytes!("../../assets/Roboto-Regular.ttf");
         let default_font = Font::try_from_bytes(font_data as &[u8]).expect("corrupted font data");
         Framebuffer {
             marker: AtomicU32::new(1),
-            device,
             frame: mem_map,
             default_font,
             var_screen_info,
             fix_screen_info,
-            swtfb_client,
+            framebuffer_update,
         }
     }
+}
 
+impl<'a> framebuffer::FramebufferBase<'a> for Framebuffer<'a> {
     fn set_epdc_access(&mut self, state: bool) {
-        if self.swtfb_client.is_some() {
-            // Not caught/handled in rm2fb => noop
-            return;
+        match &self.framebuffer_update {
+            FramebufferUpdate::Ioctl(device) => {
+                unsafe {
+                    libc::ioctl(
+                        device.as_raw_fd(),
+                        if state {
+                            MXCFB_ENABLE_EPDC_ACCESS
+                        } else {
+                            MXCFB_DISABLE_EPDC_ACCESS
+                        },
```

Comment:
We should move this outside the `unsafe` block as it's not an unsafe expression per-se also it'd collapse the ioctl call into fewer lines. (shaving that yak)